src/GeniusYield/TxBuilder/Class.hs

```diff
@@ -77,213 +95,286 @@ module GeniusYield.TxBuilder.Class
+import qualified Cardano.Api                  as Api
+import           Control.Monad.Except         (MonadError (..), liftEither)
+import qualified Control.Monad.State.Strict   as Strict
+import qualified Control.Monad.State.Lazy     as Lazy
+import qualified Control.Monad.Writer.CPS     as CPS
+import qualified Control.Monad.Writer.Strict  as Strict
+import qualified Control.Monad.Writer.Lazy    as Lazy
 import           Control.Monad.IO.Class       (MonadIO (..))
 import           Control.Monad.Random         (MonadRandom (..), RandT, lift)
 import           Control.Monad.Reader         (ReaderT)
-import           Data.List                    (nubBy)
+import           Data.Default                 (def, Default)
+import qualified Data.List.NonEmpty           as NE
 import qualified Data.Map.Strict              as Map
-import           Data.Maybe                   (listToMaybe)
+import           Data.Maybe                   (maybeToList)
 import qualified Data.Set                     as Set
 import qualified Data.Text                    as Txt
 import           Data.Time                    (diffUTCTime, getCurrentTime)
+import           Data.Word                    (Word64)
 import           GeniusYield.Imports
+import           GeniusYield.Transaction
+import           GeniusYield.TxBuilder.Common
 import           GeniusYield.TxBuilder.Errors
+import           GeniusYield.TxBuilder.Query.Class
+import           GeniusYield.TxBuilder.User
 import           GeniusYield.Types
+import           GeniusYield.Types.Key.Class  (ToShelleyWitnessSigningKey)
 import           GHC.Stack                    (withFrozenCallStack)
 import qualified PlutusLedgerApi.V1           as Plutus (Address, DatumHash,
                                                          FromData (..),
                                                          PubKeyHash, TokenName,
                                                          TxOutRef, Value)
 import qualified PlutusLedgerApi.V1.Value     as Plutus (AssetClass)
 
+-- | Class of monads for building transactions. This can be default derived if the requirements are met.
+-- Specifically, set 'TxBuilderStrategy' to 'GYCoinSelectionStrategy' if you wish to use the default in-house
+-- transaction building implementation.
+class (GYTxSpecialQueryMonad m, GYTxUserQueryMonad m, Default (TxBuilderStrategy m)) => GYTxBuilderMonad m where
+    type TxBuilderStrategy m :: Type
+    type TxBuilderStrategy m = GYCoinSelectionStrategy
+    buildTxBodyWithStrategy :: forall v. TxBuilderStrategy m -> GYTxSkeleton v -> m GYTxBody
+    default buildTxBodyWithStrategy :: forall v. (MonadRandom m, TxBuilderStrategy m ~ GYCoinSelectionStrategy)
+                                    => TxBuilderStrategy m
+                                    -> GYTxSkeleton v
+                                    -> m GYTxBody
+    buildTxBodyWithStrategy = buildTxBodyWithStrategy'
+
+buildTxBody :: GYTxBuilderMonad m => GYTxSkeleton v -> m GYTxBody
```

Comment:
Note to self: Add haddock.

Reply:
Sorry for missing this! I added it now. It's a pretty simple haddock for now but I think it suffices.